File: get_substring_indices.js
function getSubstringIndices(s, words) {
  const result = []
  const pattern = {}
  const wordLength = words[0].length

  for (const word of words) {
    pattern[word] = (pattern[word] || 0) + 1
  }

  for (let i = 0; i < wordLength; i++) {
    let pos = i
    let idxS = pos + wordLength
    let matches = {}
    let cnt = 0

    while (idxS <= s.length) {
      let word = s.slice(idxS - wordLength, idxS)

      if (pattern[word]) {
        matches[word] = (matches[word] || 0) + 1
        cnt++

        while (matches[word] > pattern[word]) {
          matches[s.slice(pos, (pos += wordLength))] -= 1
          cnt--
        }

        if (cnt === words.length) {
          result.push(pos)
        }
      } else {
        matches = {}
        cnt = 0
        pos = idxS
      }

      idxS += wordLength
    }
  }

  return result
}

const s = 'barfoofoobarthefoobarman'
const words = ['foo', 'bar', 'the']

const res = getSubstringIndices(s, words)
console.log('res::', res)

File: get_substring_indices.go
package main

import (
	"fmt"
)

func getSubstringIndices(s string, words []string) []int {
	res := []int{}
	wordLen := len(words[0])

	freqWord := make(map[string]int)

	for _, word := range words {
		if _, ok := freqWord[word]; ok {
			freqWord[word] += 1
		} else {
			freqWord[word] = 1
		}
	}

	for i := 0; i < wordLen; i++ {
		pos := i
		cnt := 0
		matches := make(map[string]int)
		idxS := pos + wordLen

		for idxS <= len(s) {
			word := s[idxS-wordLen : idxS]

			if _, ok := freqWord[word]; ok {

				if _, ok := matches[word]; ok {
					matches[word] += 1
				} else {
					matches[word] = 1
				}
				cnt += 1

				for matches[word] > freqWord[word] {
					matches[s[pos:pos+wordLen]] -= 1
					pos += wordLen
					cnt -= 1
				}

				if cnt == len(words) {
					res = append(res, pos)
				}
			} else {
				//reset
				matches = make(map[string]int)
				cnt = 0
				pos = idxS
			}
			idxS += wordLen

		}
	}

	return res
}

func main() {

	s := "barfoofoobarthefoobarman"
	words := []string{"foo", "bar", "the"}

	substringIndices := getSubstringIndices(s, words)

	fmt.Println(substringIndices)
}
